glsl.meta.slang image #3631, Review

ArielG-NV opened the pull request.

ArielG-NV committed:
[early push of code since memory qualifiers may be made into a seperate branch & pr and I rather make it simple to split the implementation if required]

all type & functions impl. for GLSL image type

added all memory qualifiers & tests for direct read/write [GLSL syntax] (DID NOT test or implement parameter qualifiers, that is next commit)
* this inlcudes emit-glsl & emit-spirv for qualifier decorations
* this also includes error handling
* this includes parsing

ArielG-NV committed:
full implementation other than Rect; all errors and basic tests are done & working

what is left:
1. need to now add Rect type support (additional TextureImpl flag)
2. tests
3. testing infrastructure to support variety of types

ArielG-NV committed:
testing framework now works with images of all types and imageBuffers -- next steps are actual tests

ArielG-NV committed:
push code for mostly working image atomics; missing int64/uint64 tests and slightly broken feature

likley due to missing code from master which I pushed for regular atomics

ArielG-NV committed:
Merge branch 'master' of https://github.com/ArielG-NV/slang into glsl.meta.slang_image_#3631

ArielG-NV committed:
fix all remaining shader image atomic issues and tests to work with float & i64/u64 fully

will now clean up code and squash the commits (since they are quite all over the place)

ArielG-NV committed:
Merge branch 'master' of https://github.com/ArielG-NV/slang into glsl.meta.slang_image_#3631

ArielG-NV committed:
refactor code to work & look correct, fix all regressions

Turned off tests for texture format R64 due to the shader use limitation of currently being only for storage buffers on most hardware (test fail cause, this is not allowed)

Changed raygen.slang & nv-ray-tracing-motion-blur.slang since both cross-compiled with glslang, which does not respect layout(rgba8) for RWBuffer's, in this scenario making the type into a SPIR-V rgba32f, which is incorrect and a known problem, this causes different code to be outputted from Slang & HLSL+GLSL->Slang paths

Clean up all code and better explain the "why" for the gimageDim definition we use various strings of Slang code, the gist is:
1. Parameters are structured as per IMAGE_PARAM keyword in spec, and we respect this in order to match specification (to allow easy code iteration)
2. sample parameters are required for functions
3. types are inconsistently named

fixed regression of breaking l-value lowering when r-value should be lowered (lower-to-ir)

fix compiler warnings

remove unneeded lambdas

`expr->type.isLeftValue = isMutableGLSLBufferBlockVarExpr(baseExpr) && (expr->type.hasReadOnlyOnTarget == false);` is an adjustment made such that a buffer block is mutable only if the block is mutable and the base expression is mutable (to handle case of readonly buffer block, immutable)

ArielG-NV committed:
remove rectangle parameter

ArielG-NV committed:
use proper const syntax and struct naming

jkwak-work commented on slang-gfx.h:
```diff
@@ -2724,5 +2732,4 @@ struct SlangSessionExtendedDesc
     uint32_t compilerOptionEntryCount = 0;
     slang::CompilerOptionEntry* compilerOptionEntries = nullptr;
 };
-
```
Please don't remove lines unnecessarily.

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -2821,6 +2821,453 @@ public vec4 shadow2DProjLod(sampler2DShadow sampler, vec4 coord, float lod)
+public const int GL_TEXTURE_CUBE_MAP_NEGATIVE_Z = 5;
+
+__generic<T : __BuiltinType>
+[ForceInline] void typeRequireChecks_image_atomic_teir1()
```
Did you mean "tier"?

jkwak-work replied:
Are you going to fix the typo?

jkwak-work commented on source/slang/slang-check-decl.cpp:
```diff
@@ -944,7 +944,27 @@ namespace Slang
+            {
+                for (auto mod : varDeclRef.getDecl()->modifiers)
+                {
+                    if (as<GLSLReadOnlyModifier>(mod))
+                    {
+                        isLValue = false;
+                        qualType.hasReadOnlyOnTarget = true;
+                        if (isLValue == false && isWriteOnly) break;
+                    }
+                    if (as<GLSLWriteOnlyModifier>(mod))
+                    {
+                        isWriteOnly = true;
+                        if (isLValue == false && isWriteOnly) break;
+                    }
```
Does this mean it is legal to have both "readonly" and "writeonly"?
Or is it an error case to have both of modifiers?

ArielG-NV replied:
it is legal to have both, it is legal to compound any number of memory quaifier's

jkwak-work replied:
The document described the case where both readonly and writeonly are used.
Can we have a test case for this case? I guess we can test if it can use "imageSize" and "length".
```
A variable could be qualified as both readonly and writeonly, disallowing both read and write.
Such variables can still be used with some queries, for example imageSize() and .length().
```

jkwak-work commented on tools/render-test/shader-input-layout.cpp:
```diff
@@ -1100,12 +1126,82 @@ namespace renderer_test
     {
         gfx::FormatInfo formatInfo;
         gfxGetFormatInfo(desc.format, &formatInfo);
-
```
I like to discourage to from making an unnecessary line removal.

jkwak-work commented on tools/render-test/shader-input-layout.cpp:
```diff
@@ -1139,12 +1235,12 @@ namespace renderer_test
                     const uint8_t* srcPixels = (const uint8_t*)srcSlice.values;
 
                     int16_t* dstPixels = (int16_t*)output.setSliceCount(i, pixelCount);
-                    
```
I like to discourage to from making an unnecessary change.

jkwak-work commented on tools/render-test/shader-input-layout.cpp:
```diff
@@ -1139,12 +1235,12 @@ namespace renderer_test
                     {
                         case 1:
                         {
-                            for (Index j = 0; j < pixelCount; ++j, srcPixels += 4, dstPixels += 1)
```
I like to discourage to from making an unnecessary change.

jkwak-work commented on tools/render-test/shader-input-layout.cpp:
```diff
@@ -1100,12 +1126,82 @@ namespace renderer_test
+                for (Index j = 0; j < pixelCount; ++j, srcPixels += 4, dstPixels++)
+                {
+                    // Copy out r
+                    dstPixels[0] = T(srcPixels[0] * (1.0f / 255));
```
I am not sure if this is a right way for integer types.
The original code looks like the assumption is that srcPixels is 8 bit format whose value was normalized.
And it is converting to a normalized floating value.
As an example, when srcPixels[0] was 254, dstPixels[0] would get 0.99f as a floating value, which makes sense.
But if the destination type is an integertype and srcPixels[0] is 254, dstPixels[0] will get 0 because int(254 / 255) is zero. The resulting value will be always zero or one, which doesn't make sense.

I think that when the destination is an integer value, you need to multiply the result with the max value of the type to keep it normalized.
Something like this,
```
dstPixels[0] = T(srcPixels[0] * limits<T>::max() / 255);
```

ArielG-NV replied:
the code assumes the type is 0 or 1, anything in-between does not matter

the code during initialization only has: one, zero, one & zero, as the initialization parameters, as a result the code works for the purposes of testing

I will comment as such

jkwak-work replied:
If the values in srcPixels are expected to be zero or one, why do we need to divide them with 255?
And is the input values never going to be changed to values other than zero and one?

The intention here looks to be converting the image format from RGBA8 to R64_UINT, as an example.
I think you need to do a proper conversion.

ArielG-NV replied:
0 or 1 was a unorm description of what it did under the-hood, I explained the process with unorm because the data is actually stored as uint8_t, 1/255*255 gives a float representation (for casting) which is how floatToHalf works. The existing system was only designed to handle 0 or 1, any other value and you are at the mercy of IEEE and floating point inaccuracies

ArielG-NV replied:
thinking about a proper conversion:

ArielG-NV replied:
lets say our input system says: load zeroPointFive, this would require the code before this which is loading the initial data to setup 128,

lets say we input two, this would require the code before this which is loading the initial data to setup 2 since 512 is not an option

jkwak-work replied:
I don't understand what you are saying.
Here is how I see it.
When we convert from RGBA8 format to RGBA32F, we do the following conversion.
```
rgba32f[0] = float(rgba8[0]) / 255.f;
```
When we convert from RGBA8 to RGBA16, as an example, we do the following,
```
rgba16[0] = rgba8[0] << 8;
```
Your implementation works for only one case where the input is 8bit integer and output is a floating point. It wouldn't work when the output is an integer format.

I don't know what you mean by 1.
It could mean a near dark color in RGBA8 format. (1/255).
Or it could mean a full white color in RGBA8 format (255/255).
Your expression is very hard to understand even with the examples.

ArielG-NV replied:
my implementation worked for all types as long as it was 0 or 1 as input. It simply took a float which was expected to be 0.0f or 1.0f, and casted it to another type.

I changed the implementation (as per the request) to work with non 0/1 inputs as well for the future.

about the new implementation: I did not want to make code that did something in the nature of: `rgba16[0] = rgba8[0] << 8;` since then when adding all 12 types we would have a big bundle of code which would be hard to iterate on, change, and reuse. I wanted to keep everything easy to understand and modify to allow easy additions of functionality. I think the new one is quite easy to work around.

jkwak-work commented on tools/gfx-unit-test/format-unit-tests.cpp:
```diff
@@ -337,6 +337,22 @@ namespace gfx_test
+                Slang::makeArray<uint32_t>(255u, 0u, 0u, 255u,
+                                           255u, 255u, 127u, 127u));
+        }
+
```
You will need to either remove this or make it working.

ArielG-NV replied:
I can't make the test working because the test uses a type that does not support uniform (non storage) use

the test is here for the same reason `auto texView = createTexView(device, size, gfx::Format::R32G32B32_FLOAT, &subData);` is added, neither are allowed to work with the resources in the test, but was added for completion(?)

jkwak-work replied:
If I were you, I would remove it because it is not hard to bring it back when we can actually test it.

If you like to keep it as it is, you will need to leave a comment and describe what needs to be done.

ArielG-NV replied:
I will remove the code, the comment would likely rot as we wait for hardware to eventually support non storage 64bit (which is an if, rather than a when).

jkwak-work reviewed (commented):
I have a few comments.
I haven't finished reviewing yet.
I will continue tomorrow.

ArielG-NV commented:
I found some things to change*, I will adjust them quickly today before more review

ArielG-NV committed:
adjust syntax

jkwak-work commented on source/slang/slang-check-modifier.cpp:
```diff
@@ -1092,15 +1094,26 @@ namespace Slang
+        }
         case ASTNodeType::GloballyCoherentModifier:
+        {
+            return (isGlobalDecl(decl) && ((as<VarDeclBase>(decl) && isGLSLInput) || as<VarDecl>(decl))) || as<ParamDecl>(decl) && isGLSLInput
```
This line is very hard to read.
And I think you are changing the behavior on HLSL side.
Before your change, as<VarDecl>(decl) had to be true to return true. But after your change it doesn't need to be.

When isGLSLInput is false, it became,
```
return (isGlobalDecl && VarDecl) || struct || interface;
```
That is different from
```
return VarDecl && (isGlobalDecl || struct || interface);
```

I will be also sensetive about the order of the evaluation.
It seems that "isGlobalDecl" is called always after "as<VarDeclBase>(decl)" or "as<VarDecl(decl)" is true; not before.

You can separate GLSL from HLSL like following:
```
if (isGLSLInput)
    return ((VarDeclBase || VarDecl) && isGlobalDecl) || ParamDecl || struct || interface;
return as<VarDecl>(decl) && (isGlobalDecl(decl) || as<StructDecl>(getParentDecl(decl)) || as<GLSLInterfaceBlockDecl>(decl));
```

OpenGL document shows two examples.
Can we have them in our test?
```
coherent buffer Block {
  readonly vec4 member1;
  vec4 member2;
};
buffer Block {
  coherent readonly vec4 member1;
  coherent vec4 member2;
};
```

ArielG-NV replied:
will adjust the decl values since they are done confusingly/wrong

but I will need to adjust it beyond what was suggested

AllowGLSL supports HLSL syntax and glsl syntax

jkwak-work commented on source/slang/slang-diagnostic-defs.h:
```diff
@@ -354,6 +356,8 @@ DIAGNOSTIC(30098, Error, nonStaticMemberFunctionNotAllowedAsDiffOperand, "non-st
 
 DIAGNOSTIC(30099, Error, sizeOfArgumentIsInvalid, "argument to sizeof is invalid")
 
+DIAGNOSTIC(30101, Error, readingFromWriteOnly, "right of '=' is writeonly (cannot be used for reading). Check modifiers.")
```
I think it will be better to be a little more generic like "Cannot read from writeonly variable."
I don't think reading operation is limited to "operator=", because the initialization expression could be in other forms.

ArielG-NV replied:
I will make it more generic, I agree with the usage.

jkwak-work commented on source/slang/slang-emit-spirv.cpp:
```diff
@@ -1895,6 +1926,7 @@ struct SPIRVEmitContext
         //
         // SPIR-V requires that the sampled/rw info on the image isn't unknown
         SLANG_ASSERT(sampled == sampledImage || sampled == readWriteImage);
+        if(isMultisampled) requireSPIRVCapability(SpvCapabilityStorageImageMultisample);
```
Can you make this two lines?
It is difficult to debug when they are all in one line.

jkwak-work commented on source/slang/slang-emit-spirv.cpp:
```diff
@@ -1765,6 +1765,37 @@ struct SPIRVEmitContext
+            requireSPIRVCapability(SpvCapabilityInt64ImageEXT);
+        default:
+            if(setCapabilityMask == SpvCapabilityStorageImageExtendedFormats) return;
+            requireSPIRVCapability(SpvCapabilityStorageImageExtendedFormats);
```
Is this intentionally fall through?
If yes, can you use [[fallthrough]];

jkwak-work commented on source/slang/slang-ir-inst-defs.h:
```diff
@@ -155,6 +150,11 @@ INST(Nop, nop, 0, 0)
+                INST(TextureShape2DType, TextureShape1DType, 0, HOISTABLE)
+                INST(TextureShape3DType, TextureShape1DType, 0, HOISTABLE)
+                INST(TextureShapeCubeType, TextureShape1DType, 0, HOISTABLE)
+                INST(TextureShapeBufferType, TextureShapeBufferType, 0, HOISTABLE)
```
Why did they become a part of TextureTypeBase?

ArielG-NV replied:
likely don't need the change I made now regardless

jkwak-work replied:
I don't know what you mean by that.
That sounds like how Yoda would say.
If the change is not needed, please don't change it.

jkwak-work reviewed (commented):
I left some comments.
I haven't read testing code yet.

ArielG-NV committed:
adjust modifier capabilitites: HLSL+GLSL --> GLSL. Notice most specifically, if the parent is a global struct we can put a memory qualifier, this does not include, struct inside a struct, with a member variable with a memory qualifier (since then you could use the struct in invalid ways). Added test for struct inside struct with member variable with memory qualifier.

adjust syntax and remove code which will rot

ArielG-NV committed:
adjust formatting for consistency

jkwak-work commented on tests/glsl-intrinsic/image/i64imageDim.slang:
```diff
@@ -0,0 +1,483 @@
+//TEST_INPUT:ubuffer(data=[0], stride=4):out,name=outputBuffer
+buffer MyBlockName2
+{
+    uint data[];
```
Shouldn't this have a size?
```
uint data[1];
```

ArielG-NV replied:
for future reference: unbound arrays are legal in glsl if final param (this is preferred to make testing modifications easier): https://www.khronos.org/opengl/wiki/Shader_Storage_Buffer_Object

I can add the 1 though

jkwak-work commented on tests/glsl-intrinsic/image/i64imageDim.slang:
```diff
@@ -0,0 +1,483 @@
+
+layout(local_size_x = 1) in;
+
+//TEST_INPUT: RWTexture1D(format=R64_SINT, size=4, content=one, mipMaps = 1):name image_1d
```
Aren't you missing "=" between "name" and "image_1d"?

ArielG-NV replied:
ironically, we do not (the code below is from the parser, note `// Optionally consume '='`):
```
                    else if (parser.AdvanceIf("name"))
                    {
                        // Optionally consume '=' 
                        if (parser.NextToken().Type == Misc::TokenType::OpAssign)
                        {
                            parser.ReadToken();
                        }

                        ioField.name = parseName(parser);
                    }
```

jkwak-work commented on tests/glsl-intrinsic/image/i64imageDim.slang:
```diff
@@ -0,0 +1,483 @@
+bool checkAllImageLoad()
+{
+    return true
+        && imageLoad(image_1d, 0).x == i64vec4(1).x
```
Why do you need to use i64vec4(1).x?
Can it be just 1LL?

ArielG-NV replied:
I can simplify these (the format was kept consistent across tests for making changing easier, but removing it is completely fine)

ArielG-NV replied:
I checked; 1 casts completely fine as well

jkwak-work commented on tests/glsl-intrinsic/image/iimageDim.slang:
```diff
@@ -0,0 +1,418 @@
+    loadCheck = loadCheck && load_image_2dMultiSampleArray() == 0;
+    resetAllImageValues();
+    return loadCheck;
+        ;
```
Probably not needed line

jkwak-work commented on tests/glsl-intrinsic/image/iimageDim.slang:
```diff
@@ -0,0 +1,418 @@
+void computeMain()
+{
+    outputBuffer.data[0] = true
+        && imageOperations()
```
I think you can just inline the function "imageOperations" here.
```
void computeMain()
{
outputBuffer.data[0] =true
        && checkAllImageSize()
        && checkAllImageLoad()
        && checkAllImageStore()
        && checkAllImageSamples()
        && checkAllImageAtomicAdd()
        && checkAllImageAtomicExchange()
        && checkAllImageAtomicMin()
        && checkAllImageAtomicMax()
        && checkAllImageAtomicAnd()
        && checkAllImageAtomicOr()
        && checkAllImageAtomicXor()
        && checkAllImageAtomicCompSwap()
        ;
}

jkwak-work commented on tests/glsl/ssboMemberMemoryQualifier.slang:
```diff
@@ -0,0 +1,48 @@
+//DIAGNOSTIC_TEST:SIMPLE(filecheck=CHECK-GLSL): -stage compute -entry computeMain -allow-glsl -target glsl
+//DIAGNOSTIC_TEST:SIMPLE(filecheck=CHECK-SPV): -stage compute -entry computeMain -allow-glsl -target spirv -emit-spirv-directly
```
What is DIAGNOSTIC_TEST?
And can we have a runtime test like this?
```
//TEST(compute, vulkan):COMPARE_COMPUTE(filecheck-buffer=BUF):-vk -compute -entry computeMain -allow-glsl
//TEST(compute, vulkan):COMPARE_COMPUTE(filecheck-buffer=BUF):-vk -compute -entry computeMain -allow-glsl -emit-spirv-directly
```
I don't know if DIAGNOSTIC_TEST does it.

ArielG-NV replied:
DIAGNOSTIC_TEST is just syntax candy. I was using it since I was basing this file off of the error check files I made which were based on a file that use DIAGNOSTIC_TEST ('unknown-opcode.slang` is an example of this).

We can have a runtime test; we don't really need it though since the goal is "make sure we can add qualifiers here" as an opposite to "make sure we cannot add qualifier here" tests. Testing if we can write 1 to a int is not really helpful since the regression is all "compiler enforcing things" and decorations which indirectly effect runtime. If glslang does not compile glslang would complain, if glslang compiles, glslang won't complain.

jkwak-work replied:
I still don't know what DIAGNSTIC_TEST is.
Can you tell me why this test uses DIAGNSTIC_TEST not TEST?

If the test is not meant for runtime result, we don't need to test it here. That is fine.

FYI, the other day, Yong told me that we cannot rely on the error messages on the downstream compilers. Slang must catch and report the syntactic errors without relying on the error messages from glslang.

ArielG-NV replied:
1. "Can you tell me why this test uses DIAGNSTIC_TEST not TEST?" because the code I based this code off of used DIAGNSTIC_TEST, they function identical, I will switch to TEST, there is no issue in doing so 

2.👍 

3. we don't rely on glslang error messages for any of the following code, I mean it as in: this is all compile time syntax candy for as far as the programmer will experience, the compiler (Slang) will error if you break syntax rules (and I am checking for these)

jkwak-work replied:
It looks like DIAGNOSTIC_TEST is meant for testing error cases.
```
struct TestOptions
{
    enum Type
    {
        Normal,             ///< A regular test
        Diagnostic,         ///< Diagnostic tests will always run (as form of failure is being tested)
    };
```
So if your intention is to testing failures or error messages, it seems appropriate to use it.

jkwak-work commented on tools/gfx/vulkan/vk-device.cpp:
```diff
@@ -1638,6 +1650,96 @@ Result DeviceImpl::createTextureResource(
+                memcpy(&data, initData->data, sizeof(data));
+                break;
+            case SLANG_SCALAR_TYPE_UINT32:
+                memcpy(&data, initData->data, sizeof(data));
```
Maybe missing a "break"

jkwak-work commented on tools/gfx/vulkan/vk-device.cpp:
```diff
@@ -1638,6 +1650,96 @@ Result DeviceImpl::createTextureResource(
+                memcpy(&data, initData->data, sizeof(data));
+            case SLANG_SCALAR_TYPE_INT64:
+            {
+                int32_t tmpData = int32_t(((int64_t*)initData->data)[0]);
```
I would prefer "reinterpret_cast" over c-style type cast.

jkwak-work commented on tools/gfx/vulkan/vk-device.cpp:
```diff
@@ -1638,6 +1650,96 @@ Result DeviceImpl::createTextureResource(
+            case SLANG_SCALAR_TYPE_FLOAT32:
+            {
+                float tmpData = float(((float*)initData->data)[0]);
+                memcpy(&data, &tmpData, sizeof(data));
```
I think it will be easier if you just assign the values without going through "data".
```
for (int i =0; i < 4; ++i)
clearColor.float32[i] = *reinterpret_cast<float*>(initData->data);
```

jkwak-work commented on tools/gfx/vulkan/vk-device.cpp:
```diff
@@ -2154,6 +2256,19 @@ Result DeviceImpl::createBufferView(
+                if (desc.type == IResourceView::Type::UnorderedAccess)
+                {
+                    bufferViewUsage.usage = VK_BUFFER_USAGE_2_STORAGE_TEXEL_BUFFER_BIT_KHR;
+                    info.pNext = &bufferViewUsage;
```
It doesn't look like you need this line.

jkwak-work reviewed (commented):
I left more comments.

ArielG-NV committed:
addressing review feedback

addressing review feedback:

change testing code to handle int and float/half correctly in all cases

adjust testing code syntax as requested

change vkdevice code to fit a different form as requested

jkwak-work commented on source/slang/glsl.meta.slang:
```diff
@@ -2821,6 +2821,453 @@ public vec4 shadow2DProjLod(sampler2DShadow sampler, vec4 coord, float lod)
+        const char *SPVSubTypePrefix;
+        const char *SPVAtomicSuffix;
+    }
+    static const kTypeWithPrefix[] =
```
Please change this to be more explicit.
```
};

static const ImageTypeInfo kTypeWithPrefix[] = 
```

jkwak-work commented on source/slang/slang-check-modifier.cpp:
```diff
@@ -1092,17 +1094,25 @@ namespace Slang
+            return as<VarDecl>(decl) && (isGlobalDecl(decl) || as<StructDecl>(getParentDecl(decl)) || as<GLSLInterfaceBlockDecl>(decl))
+                // If allowing GLSL syntax, also allow the following
+                || isGLSLInput && (as<VarDeclBase>(decl) && isGlobalDecl(decl) || as<ParamDecl>(decl) || (as<StructDecl>(getParentDecl(decl)) && isGlobalDecl(getParentDecl(decl))))
+                ;
```
Please split "isGLSLInput" into a new if-statement as I suggested.
This is still hard to read, which is easier to make a mistake.

jkwak-work commented on tools/gfx/vulkan/vk-device.cpp:
```diff
@@ -2154,6 +2257,20 @@ Result DeviceImpl::createBufferView(
+                    bufferViewUsage.usage = VK_BUFFER_USAGE_2_UNIFORM_TEXEL_BUFFER_BIT_KHR;
+                }
+                else
+                    assert(!"unhandled");
```
Please wrap the line with "{" and "}".
Depending on how "assert" macro is defined, this line could disappear and unwanted next line may become a part of "else".

ArielG-NV replied:
I had no idea assert is a macro sometimes, thank you for sharing~ I 100% should NOT be doing this with assert then...

jkwak-work commented on tools/render-test/shader-input-layout.cpp:
```diff
@@ -1100,7 +1126,109 @@ namespace renderer_test
+                out = T(127);
+                return;
+            }
+        }
```
Why is only one type treated specially?

ArielG-NV replied:
uint8_t is the src type, max size is 255.

int16, uint16, half all can safely store the type lower and upper bounds in normal c++ rules with an explicit cast or with logic (FloatToHalf).

int8_t is the only type which cannot cast 255 into itself within a strictly positive set of numbers, the spec states that bit pattern is preserved, so this would end-up being [0,127],[-1,-128]. Since testing code does not assign negatives I decided to follow this pattern and just keep things consistent with a positive only int.

ArielG-NV replied:
I can add a way to support negative fairly trivially if you prefer and believe it is the correct method

jkwak-work reviewed (commented):
I left a few comments.
We are almost getting there.

ArielG-NV committed:
adjust code as per requested for review:

1. adjusted testing code logic to handle non 0-1 values appropriately, notice int8_t will likley be the range and set order of {[0,127],[-1,-128]}, this is intentional
2. syntax adjustments for correctness
